### ISSUE TITLE
[Feature] dbt docs generate is slow against snowflake database with numerous schema (~200) and could be improved

### ISSUE THREAD

**Issue by extrospective** [dbt-labs/dbt-snowflake#1034]
### Is this a new bug in dbt-snowflake?

- [X] I believe this is a new bug in dbt-snowflake
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Our "dbt docs generate" code took 8 minutes to run every hour (against a snowflake data warehouse)
We just dropped it to 3 minutes, by dropping schemas in the same database which are unrelate to dbt.
We do not know exactly but this may have been a drop from 300-400 schemas to 200 schemas.

This discovery allowed us to see that either:
a. dbt could write its query of information_schema in a different way (more attuned to performance issues) or
b. dbt labs could advise users not to have any unrelated schema in the same database (thereby improving performance with little effort by dbt labs)

I also suggest:
c. standard instrumentation in the debug log to call out this long running query of snowflake information_schema for transparency
(We were able to see the query with --debug, but I think it's going to continue to be an issue so inclusion in a standard log will raise awareness, leading to more workarounds and design changes).

Background: why do we have many schema on the same server?
* we are generating schema for branch PRs - not to test documentation - and just as a convenient location from an access control standpoint.

What is the block of code which is underperforming? Queries of the form:
```
select yyy from xxx.INFORMATION_SCHEMA.columns
```

Now in our case we observe that a simple select count(*) from this virtual table (INFORMATION_SCHEMA.columns) takes 1:45.

So it seems obvious to me that reduction of schema count will be crucial for optimal performance - and we will do that.  (And it could be related to raw column count rather than schema count - we do not know yet.  We have about 890K columns in our database across all schemas.)

The query that dbt employs does not fully use guidance from snowflake on use of this table, and so there may be room for improvement (a hypothesis).  Links below.

As an example of improvement potential, a snippet from the long query reads:
```
    from dbt_db.INFORMATION_SCHEMA.tables
            where (upper("table_schema") = upper('dbt_mandersen') or upper("table_schema") = upper('dbt_mandersen_ds_publish') or upper("table_schema") = upper('dbt_mandersen_internal_data') or upper("table_schema") = upper('dbt_mandersen_snapshots') or upper("table_schema") = upper('dbt_mandersen_vice') or upper("table_schema") = upper('dbt_mandersen_ironzion') or upper("table_schema") = upper('dbt_mandersen_ai_ingest'))
        ),
```
these upper(x) statements reduce the ability of the database to leverage indices.

Here are some related articles on using the information schema in snowflake:
* https://stackoverflow.com/questions/65041926/why-are-information-schema-queries-so-slow-in-snowflake-is-there-any-ways-to-sp
* https://community.snowflake.com/s/article/metadata-operations-throttling#:%7E:text=What%20are%20the%20Metadata%20Operations,warehouse%20or%20customer%20data%20storage.

Again I do not know whether dbt labs will make big strides here, and if it is impossible, simply communicating this as a key aspect driving "dbt docs generate" running long in snowflake may help users of dbt.

### Expected Behavior

Expect "dbt docs generate" to consistently run in under one minute, with much parallelization

### Steps To Reproduce

snowflake on us-east

dbt version 1.7

### Relevant log output

```shell
We ran "dbt docs generate" with --debug switch.
And then saw the queries of the information_schema.tables etc.

I do not want to include the full query here for privacy reasons.

That query was taking a long time when we just popped it out to run it in snowflake ourselves.  It is the primary cause of "dbt docs generate" being slow against snowflake.
```

### Environment

```markdown
- OS:
- Python:
- dbt-core:
- dbt-snowflake:
```

**Comment by dataders** [comment]
> 1.  dbt could write its query of information_schema in a different way (more attuned to performance issues)

100% agree! in fact, #1035 was opened only one hour before this one, and we've just merged the associated PR: #982

> 2. dbt labs could advise users not to have any unrelated schema in the same database (thereby improving performance with little effort by dbt labs)

also agree! I just opened https://github.com/dbt-labs/docs.getdbt.com/issues/5502 to improve this, though I'd love to get ideas on where we might best call that out on our docs site. plenty of options

> 3. standard instrumentation in the debug log to call out this long running query of snowflake information_schema for transparency

this one I'm not so clear about. you're proposing that something that shows up in `debug` mode should show up all the time? Like the `dbt run` should report everytime that `list_schemas()` macro is invoked along with timing information? This seems like great information for `debug`. Please tell me more!

**Comment by extrospective** [comment]
Regarding 3:

I was assuming you would not be able to get the information schema query down to just a few seconds.
If you were to do that, then I think you can disregard this suggestion.  The linked tickets suggest significant speedup is possible.

If I run "dbt docs generate" there is little useful information.  If this were to continue to be a multi-minute query some info could be added there.

When you run "dbt docs generate --debug" - there is a line which reads for me:
```
SQL status: SUCCESS 24005 in 239.0 seconds
```

That line is perhaps the most important in the file - since it accounts for much of the time.

One option - make the line more clear what it refers to.
3 seconds later in the log file is what presumably is the connection which caused it:
```
13:32:56  SQL status: SUCCESS 24005 in 239.0 seconds
13:32:59  On dbt_db.information_schema: Close
```
if there is parallelization, no guarantee something will not be in between - so I might like a more descriptive first line.

And if the problem is not made to "go away", I could imagine accumulating some critical timing information for these important queries, and at the end of the log file process (with --debug) flushing that critical information there - so it stands out from the 43K line file.
